tritium-caffeine/src/main/java/com/palantir/tritium/metrics/caffeine/CacheStats.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics.caffeine;
+
+import com.codahale.metrics.Counter;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.github.benmanes.caffeine.cache.stats.StatsCounter;
+import com.google.common.collect.ImmutableMap;
+import com.palantir.tritium.metrics.cache.CacheMetrics;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.util.Arrays;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import org.checkerframework.checker.index.qual.NonNegative;
+
+public final class CacheStats implements StatsCounter, Supplier<StatsCounter> {
+    private final String name;
+    private final Counter hitCounter;
+    private final Counter missCounter;
+
+    private final LongAdder totalLoadNanos = new LongAdder();
+    private final Counter loadSuccessCounter;
+    private final Counter loadFailureCounter;
+    private final Counter evictionsTotalCounter;
+    private final ImmutableMap<RemovalCause, Counter> evictionCounters;
+
+    /**
+     * Creates a {@link CacheStats} instance that registers metrics for Caffeine cache statistics.
+     * <p>
+     * Example method
+     * <pre>
+     *         CacheStats cacheStats = CacheStats.of(taggedMetricRegistry, "your-cache-name");
+     *         LoadingCache&lt;Integer, String> cache = cacheStats.register(Caffeine.newBuilder()
+     *                 .recordStats(cacheStats.record())
+     *                 .build(key -> computeSomethingExpensive(key));
+     * </pre>
+     * @param taggedMetricRegistry tagged metric registry to add cache metrics
+     * @param name cache name
+     * @return Caffeine stats instance to register via
+     * {@link com.github.benmanes.caffeine.cache.Caffeine#recordStats(Supplier)}.
+     */
+    public static CacheStats of(TaggedMetricRegistry taggedMetricRegistry, String name) {
+        return new CacheStats(CacheMetrics.of(taggedMetricRegistry), name);
+    }
+
+    public static CacheStats of(CacheMetrics metrics, String name) {
+        return new CacheStats(metrics, name);
+    }
```

Comment:
I'd prefer if we made `CacheMetrics` private -- metric-schema doesn't make any guarantees around backward compatibility of generated code because it's expected to be consumed by the project which declares the metrics.

I suppose we may want to reuse the same metrics for guava and caffeine caches. If that's the case, I might recommend a new jar module with `internal` in the name, like `tritium-cache-internal`, where metrics may be defined and generated as public classes. The `tritium-cache-internal` module can be used as an `implementation` dep into the tritium modules which instrument specific cache implementations, so that the classes aren't directly visible to consumers.

Reply:
Makes sense, I can move the metric-schema definition and make package private. 

Generally I'd like to deprecate all uses of Guava caches and migrate any remaining to Caffeine, so not sure its worth adding similar for Guava caches. We already have [`BanGuavaCaches` in baseline](https://github.com/palantir/gradle-baseline/blob/a40f0782f0a1913748408f33eeceec0cdba99d92/.baseline/checkstyle/checkstyle.xml#L177-L181)

Reply:
updated